### ISSUE TITLE
can we collate data and save to disk when overriding __inc__?

### ISSUE THREAD

**Issue by zhan-xu** [rusty1s/pytorch_geometric#1545]
## ❓ Questions & Help

Hi, thanks for the great library! I have a question. I am making customized data class just like the BipartiteData you mentioned here:https://pytorch-geometric.readthedocs.io/en/latest/notes/batching.html. However, I notice in the examples you didn't call:
```
data, slices = self.collate(data_list)
torch.save((data, slices), self.processed_paths[0])
```
Since my dataset is large, I prefer to make it once and save it. However, I failed to run self.collate.  My customized Data class is like this
```
class CorrData(Data):
    def __init__(self, vtx, pts, edge_vtx, edge_pts, corr, name):
        super(CorrData, self).__init__()
        self.vtx = vtx
        self.pts = pts
        self.edge_index_vtx = edge_vtx
        self.edge_index_pts = edge_pts
        self.corr = corr
        self.name = name

    def __inc__(self, key, value):
        if key == 'edge_index_vtx':
            return self.vtx.size(0)
        if key == 'edge_index_pts':
            return self.pts.size(0)
        if key == 'corr':
            return torch.tensor([[self.vtx.size(0)], [self.pts.size(0)]])
        else:
            return super(CorrData, self).__inc__(key, value)

    def __cat_dim__(self, key, value):
        if 'edge_index' in key:
            return 1
        else:
            return 0
```
The reported error is like this:

> File "xxxx/datasets/corr_dataset.py", line 141, in process
>     data, slices = self.collate(data_list)
>   File "xxxx/torch_geometric/data/in_memory_dataset.py", line 93, in collate
>     data = data_list[0].__class__()
> TypeError: __init__() missing 6 required positional arguments: 'vtx', 'pts', 'edge_vtx', 'edge_pts', 'corr', and 'name'

Any suggestion about how to solve this? Thanks!

**Comment by rusty1s** [comment]
In general, `collate` should be usable even for custom `Data` objects. You can fix the error by changing:
```python
def __init__(self, vtx, pts, edge_vtx, edge_pts, corr, name):
```
to
```python
def __init__(self, vtx=None, pts=None, edge_vtx=None, edge_pts=None, corr=None, name=None):
```

**Comment by zhan-xu** [comment]
Thanks for your solution. It solves my problem.

However, I met another problem. When I try to iterate the dataset with 

```
for data in train_loader:
```
It throws an error:

>   File "xxx/torch/utils/data/dataloader.py", line 345, in __next__
>     data = self._next_data()
>   File "xxx/torch/utils/data/dataloader.py", line 385, in _next_data
>     data = self._dataset_fetcher.fetch(index)  # may raise StopIteration
>   File "xxx/torch/utils/data/_utils/fetch.py", line 47, in fetch
>     return self.collate_fn(data)
>   File "xxx/torch_geometric/data/dataloader.py", line 34, in __call__
>     return self.collate(batch)
>   File "xxx/torch_geometric/data/dataloader.py", line 15, in collate
>     return Batch.from_data_list(batch, self.follow_batch)
>   File "xxx/torch_geometric/data/batch.py", line 53, in from_data_list
>     item = item + cum if cum != 0 else item
> RuntimeError: bool value of Tensor with more than one value is ambiguous

I tried the examples on https://pytorch-geometric.readthedocs.io/en/latest/notes/batching.html#pairs-of-graphs. The "Pairs of Graphs" example works well, but the "Bipartite Graphs" example has the same error.

I have tried with both pytorch_geometric 1.5.0 and 1.6.0 (installed from pip). I gave it a deeper look, and seems in torch_geometric/data/batch.py, line 53
```
item = item + cum if cum != 0 else item
```
This line crashes when cum is a tensor with multiple element. To walk around, I temporarily changed it into
```
item = item + cum
```
After this, both "Bipartite Graphs" and my code can run, but throw out a warning:

> WARNING:root:The number of nodes in your data object can only be inferred by its edge indices, and hence may result in unexpected batch-wise behavior, e.g., in case there exists isolated nodes. Please consider explicitly setting the number of nodes for this data object by assigning it to data.num_nodes

The warning makes sense to me. However, like in the case of "Bipartite Graphs", a pair of graphs are in the same data sample, we may need to explicitly define num_nodes for both x_s and x_t? Any suggestion about this?

Thanks for your time.

**Comment by rusty1s** [comment]
Regarding your first issue: This was a bug introduced in PyG 1.6.0, but it is fixed in PyG 1.6.1. Sorry!
Regarding your warning: Do you know where exactly this warning does occur? This shouldn't happen either.

**Comment by zhan-xu** [comment]
Seems the warning is gone for now. Now sure what I did. I will mark this as thread as close for now. Thank you for all the help here.
